src/gateway_gatt_char_wifi_ssid.erl

```diff
@@ -51,53 +52,96 @@ read_value(State=#state{value=Value}, _) ->
+        _ ->
+            {noreply, State}
+    end;
+handle_signal(_SignalID, Msg, State=#state{}) ->
+    %% State#state.service_signal /= _SignalID.
+    %% Signal from a different service
+    Args = ebus_message:args(Msg),
+    lager:warning("unexpected service_signal: ~p", [Args]),
+    case Args of
+        {ok, ["State", "online"]} ->
+            NextState = maybe_notify_value(update_value(State, "disconnect")),
+            {noreply, maybe_notify_value(update_value(NextState, "online"))};
         _ ->
             {noreply, State}
     end.
 
-
 %%
 %% Internal
 %%
 
--spec update_value(#state{}) -> #state{}.
-update_value(State=#state{service_path=CurrentPath}) ->
+update_value(State=#state{service_path=CurrentPath}, ServiceState) ->
     %% Start a scan to repopulate list of visible wifi services
     connman:scan(wifi),
-    case {CurrentPath, gateway_config:wifi_services_online()} of
+    case ServiceState of
+        "online" ->
+            on_online_state(CurrentPath, State);
+        _ ->
+            on_offline_states(CurrentPath, State)
+    end.
+
+on_online_state(CurrentPath, State) ->
+    WiFiServicesOnline = gateway_config:wifi_services_online(),
+    case {CurrentPath, WiFiServicesOnline} of
         {CurrentPath, [{_, CurrentPath} | _]} ->
             %% No change in path
             lager:info("WiFi unchanged for ~p at: ~p", [State#state.value, CurrentPath]),
             State;
-        {undefined, []} ->
-            %% No path and not connected
-            lager:info("WiFi still disconnected"),
-            State;
         {undefined, [{Value, NewPath} | _]} ->
             %% From no path to a path
             lager:info("WiFi connected to: ~p at ~p", [Value, NewPath]),
             {ok, SignalID} = connman:register_state_notify({path, NewPath}, self(), State#state.path),
             State#state{value=list_to_binary(Value), service_path=NewPath, service_signal=SignalID};
-        {CurrentPath, []} ->
-            %% Disconnected from wifi
-            lager:info("WiFi disconnected from: ~p", [State#state.value]),
-            connman:unregister_state_notify(State#state.service_signal, self(), State#state.path),
-            State#state{service_signal=undefined, value= <<"">>, service_path=undefined};
         {CurrentPath, [{Value, NewPath} | _]} ->
             %% Switching paths
             lager:info("WiFi changed from ~p to: ~p", [State#state.value, Value]),
             connman:unregister_state_notify(State#state.service_signal, self(), State#state.path),
             {ok, SignalID} = connman:register_state_notify({path, NewPath}, self(), State#state.path),
-            State#state{value=list_to_binary(Value), service_path=NewPath, service_signal=SignalID}
+            State#state{value=list_to_binary(Value), service_path=NewPath, service_signal=SignalID};
+        _ ->
+            lager:warning("early WiFi online event"),
+            State
+    end.
+
+on_offline_states(CurrentPath, State) ->
```

Comment:
Rename this to something closer to what it actually does? Looks like it `maybe_unregister_notify`?

Reply:
Done.